lib/AL_USDMaya/AL/usdmaya/Global.cpp

```diff
@@ -45,12 +58,165 @@ namespace {
   std::atomic<size_t> readDepth;
+
+#if defined(WANT_UFE_BUILD)
+  // The proxy shape has an internal cache which needs to update when any of
+  // its UFE scene items are selected and transformed.
+  class UfeTransformObserver : public Ufe::Observer
+  {
+  public:
+    UfeTransformObserver() : Ufe::Observer()
+    {
+    }
+
+    ~UfeTransformObserver() override
+    {
+    }
+
+    void operator()(const Ufe::Notification& notification) override
+    {
+      auto xformChanged = dynamic_cast<const Ufe::Transform3dChanged*>(&notification);
+      if (xformChanged == nullptr) return;
+
+      // Action for USD scene items only.
+      Ufe::SceneItem::Ptr sceneItem = xformChanged->item();
+      if (!sceneItem || (sceneItem->runTimeId() != AL::usdmaya::USD_UFE_RUNTIME_ID)) return;
+
+      std::string mayaPath = sceneItem->path().popSegment().popHead().string();
+
+      MSelectionList sl;
+      sl.add(MString(mayaPath.c_str()));
+
+      MObject object;
+      MStatus status = sl.getDependNode(0, object);
+      if (!status) return;
+
+      MFnDependencyNode dependNode(object, &status);
+      if (!status) return;
+
+      if (dependNode.typeId() == AL::usdmaya::nodes::ProxyShape::kTypeId)
+      {
+        auto proxyShape =
+          dynamic_cast<AL::usdmaya::nodes::ProxyShape*>(dependNode.userNode());
+        if (proxyShape)
+        {
+          proxyShape->clearBoundingBoxCache();
+        }
+      }
+    }
+  };
+#endif
 }
 
 namespace AL {
 namespace usdmaya {
 
 
+//----------------------------------------------------------------------------------------------------------------------
+#if defined(WANT_UFE_BUILD)
+  // Observe UFE scene items for transformation changed only when they are selected.
+  class Global::UfeSelectionObserver : public Ufe::Observer
+  {
+  public:
+    UfeSelectionObserver()
+      : Ufe::Observer()
+      , m_ufeTransformObserver(std::make_shared<UfeTransformObserver>())
+      , m_openingFile(false)
+    {
+    }
+
+    ~UfeSelectionObserver() override
+    {
+      clear();
+    }
+
+    void clear()
+    {
+      for (auto si : m_observedSceneItems)
+      {
+        Ufe::Transform3d::removeObserver(si, m_ufeTransformObserver);
+      }
+
+      m_observedSceneItems.clear();
+    }
+
+    void observe(const Ufe::SceneItem::Ptr& si)
+    {
+      if (si &&
+        (si->runTimeId() == USD_UFE_RUNTIME_ID) &&
+        Ufe::Transform3d::addObserver(si, m_ufeTransformObserver))
+      {
+        m_observedSceneItems.push_back(si);
+      }
+    }
+
+    void openingFile(bool val)
+    {
+      m_openingFile = val;
+    }
+
+    void operator()(const Ufe::Notification& notification) override
+    {
+      // During Maya file read, each node will be selected in turn, so we get
+      // notified for each node in the scene.  Prune this out.
+      if (m_openingFile)
+      {
+        return;
+      }
+
+      auto selectionChanged = dynamic_cast<const Ufe::SelectionChanged*>(&notification);
```

Comment:
ewwww :(

Reply:
We discussed this internally and agree this is not optimal. In order to not make too many changes to this pull request, we've logged a JIRA ticket (MAYA-95882 - AL_USDMaya plugin : Optimize selection notifications) for ourselves to clean this up.